akka-contrib/src/main/scala/akka/contrib/pattern/TransparentExponentialBackoffSupervisor.scala

```diff
@@ -0,0 +1,192 @@
+  maxBackoff: FiniteDuration,
+  randomFactor: Double)
+  extends Actor
+  with Stash
+  with ActorLogging {
+
+  import TransparentExponentialBackoffSupervisor._
+  import context._
+
+  override val supervisorStrategy = OneForOneStrategy(maxNrOfRetries = -1) {
+    case ex ⇒
+      val defaultDirective: Directive =
+        super.supervisorStrategy.decider.applyOrElse(ex, (_: Any) ⇒ Escalate)
+      val maybeDirective: Option[Directive] = decider
+        .map(_.applyOrElse(ex, (_: Any) ⇒ defaultDirective))
+
+      // Get the directive from the specified decider or fallback to
+      // the default decider.
+      // Whatever the final Directive is, we will translate all Restarts
+      // to our own Restarts, which involves stopping the child.
+      maybeDirective.getOrElse(defaultDirective) match {
+        case Restart ⇒
+          val childRef = sender
+          become({
+            case Terminated(`childRef`) ⇒
+              unwatch(childRef)
+              unbecome()
+              self ! ScheduleRestart(childRef)
+            case Terminated(other) ⇒
+              log.debug(s"Received unexpected Terminated message from: $other")
+            case _ ⇒
+              stash()
+          }, discardOld = false)
+          Stop
+        case other ⇒ other
+      }
+  }
+
+  // Initialize by starting up and watching the child
+  self ! StartChild
+
+  def receive = waitingToStart(-1, false)
+
+  def waitingToStart(numRestarts: Int, scheduleCounterReset: Boolean): Receive = {
+    case StartChild ⇒
+      val childRef = actorOf(props)
+      watch(childRef)
+      unstashAll()
+      if (scheduleCounterReset) {
+        system.scheduler.scheduleOnce(minBackoff, self, ResetRestartCount(numRestarts + 1))
+      }
+      become(watching(childRef, numRestarts + 1))
+    case _ ⇒ stash()
+  }
+
+  // Steady state
+  def watching(childRef: ActorRef, numRestarts: Int): Receive = {
+    case ScheduleRestart(`childRef`) ⇒
+      val delay = akka.pattern.BackoffSupervisor.calculateDelay(
+        numRestarts, minBackoff, maxBackoff, randomFactor)
+      system.scheduler.scheduleOnce(delay, self, StartChild)
+      become(waitingToStart(numRestarts, true))
+      log.info(s"Restarting child in: $delay; numRestarts: $numRestarts")
+    case ScheduleRestart(other) ⇒
+      log.debug(s"Ignoring unexpected ScheduleRestart message from: $other")
+    case ResetRestartCount(last) ⇒
+      if (last == numRestarts) {
+        log.debug(s"Last restart count [$last] matches current count; resetting")
+        become(watching(childRef, 0))
+      } else {
+        log.debug(s"Last restart count [$last] does not match the current count [$numRestarts]")
+      }
+    case Terminated(`childRef`) ⇒
+      log.debug(s"Terminating, because child [$childRef] terminated itself")
+      stop(self)
+    case Terminated(other) ⇒
+      log.debug(s"Ignoring unexpected Terminated message from: $other")
+    case msg ⇒
+      childRef.forward(msg)
```

Comment:
another thing that was pointed out in the ticket is that we should support the `Passivate` message for cluster sharding. That is sent to the parent. I think we can forward all messages from the child (`sender()`) to the parent.